sysml.library/Kernel Library/SequenceFunctions.kerml

```diff
@@ -0,0 +1,48 @@
+package SequenceFunctions {
+	private import Base::Anything;
+	private import ScalarValues::*;
+	private import ControlFunctions::exists;
+	private import ControlFunctions::select;
+	private import ControlFunctions::reject;
+
+	function size(seq: Anything[0..*] ordered nonunique): Natural {
+		isEmpty(seq)? 0: size(tail(seq)) + 1
+	}
+	function isEmpty(seq: Anything[0..*] ordered nonunique): Boolean {
+		seq == null
+	}
+	function notEmpty(seq: Anything[0..*] ordered nonunique): Boolean {
+		!isEmpty(seq)
+	}
+	function includes(seq: Anything[0..*] ordered nonunique, value: Anything[1]): Boolean {
+		seq->exists {in x; x == value}
+	}
+	function excludes(seq: Anything[0..*] ordered nonunique, value: Anything[1]): Boolean {
+		!seq->includes(value)
+	}
+	
+	function union(seq1: Anything[0..*] ordered nonunique, seq2: Anything[0..*] ordered nonunique): Anything[0..*] ordered nonunique {
+		(seq1, seq2)
+	}
+	function intersection(seq1: Anything[0..*] ordered nonunique, seq2: Anything[0..*] ordered nonunique): Anything[0..*] ordered nonunique {
+		seq1->select {in x; seq2->includes(x)}
+	}
+	function including(seq: Anything[0..*] ordered nonunique, value : Anything[1]): Anything[0..*] ordered nonunique {
+		(seq, value)
+	}
+	function excluding(seq: Anything[0..*] ordered nonunique, value : Anything[1]): Anything[0..*] ordered nonunique {
+		seq->reject {in x; x == value}
+	}
+	
+	function head(seq: Anything[0..*] ordered nonunique): Anything[0..1] {
+		seq[1]
+	}
+	function tail(seq: Anything[0..*] ordered nonunique): Anything[0..*] ordered nonunique {
+		seq->excluding(head(seq))
```

Comment:
It's interesting that the definition of seq seems LISP-like list, I mean consisting car (head) and cdr (tail).  However, the definition of tail looks awkward because it seems that excluding all of the elements that "equal" to the head.  So  what's the definition of "==" operator?  Is it "eq" or "equals"?  If it's "eq", does KerML or SysML have equal-like operator? Also the definition of union is interesting.  It seems to create a new sequence consisting of seq1 and seq2 rather than coalescing these two sequences.  If it's correct, how can we write to build a sequence of these two?

Reply:
1. You are right, the definition of `tail` is incorrect. I will fix it.
2. The intent is that `==` can operate on sequences, in which case they are equal if they are the same length and the elements in the same position are equal. The equality of individual elements is then given by different specializes of `==` depending on the type of those elements. Really, though, I think the equality operation for individual values needs to be different from the equality operation over sequence, or it is hard to avoid a circular definition, since any single value is a sequence of length 1.
3. "Sequences" here are just features that have multiple values with some multiplicity. There is no "sequence" value itself, so all such sequences are "flat, and it is not possible to have a sequence as a member of a sequence. So, the `,` operator is actually equivalent to `union` -- it concatenates sequences. This is just the semantics of multiplicity that we have carried over from UML. If you want to have collections as values, allowing collection values in collections, then you need to use the types from the `Collections` package.